src/hypothesis/internal/conjecture/minimizer.py

```diff
@@ -78,55 +86,95 @@ def _shrink_index(self, i, c):
+                for c in ss:
+                    if self._shrink_index(i, c):
+                        for c in hrange(self.current[i]):
+                            if c in ss:
+                                continue
+                            if self._shrink_index(i, c):
+                                break
+                        break
+
     def run(self):
         if not any(self.current):
             return
+        if len(self.current) == 1:
+            for c in range(self.current[0]):
+                if self.incorporate(hbytes([c])):
+                    break
+            return
         if self.incorporate(hbytes(self.size)):
             return
+        if self.incorporate(hbytes([0] * (self.size - 1) + [1])):
+            return
         change_counter = -1
         while self.current and change_counter < self.changes:
             change_counter = self.changes
-            for c in hrange(max(self.current)):
-                if self.incorporate(
-                    hbytes(min(b, c) for b in self.current)
-                ):
-                    break
 
-            for c in sorted(set(self.current), reverse=True):
-                for d in hrange(c):
-                    if self.incorporate(
-                        hbytes(d if b == c else b for b in self.current)
-                    ):
-                        break
+            self.ddlower(0)
+            self.ddlower(1)
+            self.ddshift()
+            self.ddsub()
 
-            for c in hrange(max(self.current)):
-                k = len(self.current) // 2
-                while k > 0:
-                    i = 0
-                    while i + k <= len(self.current):
-                        self.incorporate(
-                            self.current[:i] +
-                            hbytes(min(b, c) for b in self.current[i:i + k]) +
-                            self.current[i + k:]
-                        )
-                        i += k
-                    k //= 2
+            if change_counter != self.changes:
+                continue
+
+            for c in sorted(set(self.current) - {0, 1, max(self.current)}):
+                self.ddlower(c)
+
+            if change_counter != self.changes:
+                continue
+            self.shrink_indices()
 
             if change_counter != self.changes or self.cautious:
                 continue
 
-            for i in hrange(self.size):
-                t = self.current[i]
-                if t > 0:
-                    ss = small_shrinks[self.current[i]]
-                    for c in ss:
-                        if self._shrink_index(i, c):
-                            for c in hrange(self.current[i]):
-                                if c in ss:
-                                    continue
-                                if self._shrink_index(i, c):
-                                    break
-                            break
+            self.rotate_suffixes()
```

Comment:
I think the reasoning was that bulk operations should all happen before shrink\_indices, which is pretty expensive, but I don't remember if this reasoning was backed by benchmarking.